### ISSUE TITLE
Why model size doesn't decrease after INT8 Quantization using pytorch-quantization-tool?

### ISSUE THREAD

**Issue by Water2style** [NVIDIA/TensorRT#1690]
Why model size doesn't decrease after INT8 Quantization using pytorch-quantization-tool?

I use post-training-quantization follow https://docs.nvidia.com/deeplearning/tensorrt/pytorch-quantization-toolkit/docs/tutorials/quant_resnet50.html

After saving the model, I find the model size is similar to original one, after export to ONNX , the size is same.
Why not the model size decrease to 1/4 compared with the original one?
How could I see the real size of the INT8 model?

Thanks!

**Comment by zerollzeng** [comment]
the onnx onnx model still contains FP32 activation and weights, the toolkit adds Q/DQ layers which contains the scale, you can see the Q/DQ layers in Netron.

**Comment by Water2style** [comment]
> the onnx onnx model still contains FP32 activation and weights, the toolkit adds Q/DQ layers which contains the scale, you can see the Q/DQ layers in Netron.

Thank you.
I have two more questions,
1)do you how can I check the reduced model size after quantitazion ? 
2) I use Netron check the model, I find the weighs in "conv" op is still FP32 data, where is the INT8 data ? And what is the Q/DQ layers for?

**Comment by zerollzeng** [comment]
> 1)do you how can I check the reduced model size after quantitazion ?
> 2) I use Netron check the model, I find the weighs in "conv" op is still FP32 data, where is the INT8 data ? And what is the Q/DQ layers for?

1. You can save the TensorRT engine and check its size, ideally, it should be about 1/4 of the original size. eg. /usr/src/tensorrt/bin/trtexec --onnx=/usr/src/tensorrt/data/resnet50/ResNet50.onnx --int8 --fp16 --saveEngine=/tmp/resnet50.plan
2. Q/DQ layer is the scale that is used for quantizing to INT8/de-quantizing back to FP32. the weight and activation in ONNX is FP32, they will Quantize by Q layer then do the conv and de-quantize back to FP32 by DQ layer. eg. input(FP32) -> Q -> input(IN8) -> Conv(INT8 precision) -> output(INT8) -> DQ -> output(FP32)

**Comment by Water2style** [comment]
> > 1)do you how can I check the reduced model size after quantitazion ?
> > 2) I use Netron check the model, I find the weighs in "conv" op is still FP32 data, where is the INT8 data ? And what is the Q/DQ layers for?
> 
> 1. You can save the TensorRT engine and check its size, ideally, it should be about 1/4 of the original size. eg. /usr/src/tensorrt/bin/trtexec --onnx=/usr/src/tensorrt/data/resnet50/ResNet50.onnx --int8 --fp16 --saveEngine=/tmp/resnet50.plan
> 2. Q/DQ layer is the scale that is used for quantizing to INT8/de-quantizing back to FP32. the weight and activation in ONNX is FP32, they will Quantize by Q layer then do the conv and de-quantize back to FP32 by DQ layer. eg. input(FP32) -> Q -> input(IN8) -> Conv(INT8 precision) -> output(INT8) -> DQ -> output(FP32)

Thank you very much ! I think I got the basic mechanisim from your answer!

**Comment by tanmay2798** [comment]
I am trying to quantize a pytorch model having layers (Conv1d and linear). The issues are - 
1. Quantized model inference in pytorch is slower than original fp32 model. (7 times)
2. Quantized model size is bigger than fp32 model (136 kb vs 11 kb)
3. When I try to access the weights of Quantized model by going in state_dict and check the dtype of weights, it is still fp32. Why is it so?
m.cpu().state_dict()['model.0.weight'][0][0][0].numpy() - array(0.10222942, dtype=float32)  [where m is quantized model]

Quantized Model - 
CNN1dCCF(
  (model): Sequential(
    (0): QuantConv1d(
      1, 8, kernel_size=(15,), stride=(2,), padding=(1,)
      (_input_quantizer): TensorQuantizer(8bit fake per-tensor amax=12.9095 calibrator=HistogramCalibrator scale=1.0 quant)
      (_weight_quantizer): TensorQuantizer(8bit fake axis=0 amax=[0.2106, 0.3614](8) calibrator=MaxCalibrator scale=1.0 quant)
    )
    (1): BatchNorm1d(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU()
    (3): QuantConv1d(
      8, 8, kernel_size=(10,), stride=(2,), padding=(1,)
      (_input_quantizer): TensorQuantizer(8bit fake per-tensor amax=10.8750 calibrator=HistogramCalibrator scale=1.0 quant)
      (_weight_quantizer): TensorQuantizer(8bit fake axis=0 amax=[0.2187, 0.2986](8) calibrator=MaxCalibrator scale=1.0 quant)
    )
    (4): BatchNorm1d(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (5): ReLU()
    (6): QuantAdaptiveAvgPool1d(
      output_size=1
      (_input_quantizer): TensorQuantizer(8bit fake per-tensor amax=6.7122 calibrator=MaxCalibrator scale=1.0 quant)
    )
    (7): Flatten(start_dim=1, end_dim=-1)
    (8): Dropout(p=0.25, inplace=False)
    (9): QuantLinear(
      in_features=8, out_features=1, bias=True
      (_input_quantizer): TensorQuantizer(8bit fake per-tensor amax=3.3145 calibrator=HistogramCalibrator scale=1.0 quant)
      (_weight_quantizer): TensorQuantizer(8bit fake axis=0 amax=1.8564 calibrator=MaxCalibrator scale=1.0 quant)
    )
  )
)

**Comment by jinfagang** [comment]
@Water2style This is what I didn't expected. Why such design?? ONNXRuntime inside onnx quantizer can also generated smaller size of onnx. why didn't you do that? 

Meanwhile, I also want ask, how can use exported quantized onnx model inference via other framework, etc ncnn?

**Comment by jackwish** [comment]
Thanks @zerollzeng for your support!
@Water2style Is this still an issue for you?

**Comment by nvpohanh** [comment]
Closing this for now due to >14 days with no response. Please feel free to reopen if the issue still exists. Thanks!

> Why such design?? ONNXRuntime inside onnx quantizer can also generated smaller size of onnx. why didn't you do that?

@jinfagang This is not an issue in TRT, but in ONNX, which we don't have control of. As Zero said, if you build a TRT engine with INT8 enabled, the engine size should be roughly 1/4 of the original ONNX file.
